src/libraries/Common/src/SourceGenerators/CSharpSyntaxUtilities.cs

```diff
@@ -0,0 +1,68 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Globalization;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+
+namespace SourceGenerators;
+
+internal static class CSharpSyntaxUtilities
+{
+    // Standard format for double and single on non-inbox frameworks.
+    public const string DoubleFormatString = "G17";
+    public const string SingleFormatString = "G9";
+
+    // Format a literal in C# format -- works around https://github.com/dotnet/roslyn/issues/58705
+    public static string FormatLiteral(object? value, TypeRef type)
+    {
+        if (value == null)
+        {
+            return $"default({type.FullyQualifiedName})";
+        }
+
+        if (type.TypeKind is TypeKind.Enum)
+        {
+            // Return the numeric value.
+            return FormatNumber();
+        }
+
+        switch (value)
+        {
+            case string @string:
+                return SymbolDisplay.FormatLiteral(@string, quote: true); ;
+            case char @char:
+                return SymbolDisplay.FormatLiteral(@char, quote: true);
+            case double.NegativeInfinity:
+                return "double.NegativeInfinity";
+            case double.PositiveInfinity:
+                return "double.PositiveInfinity";
+            case double.NaN:
+                return "double.NaN";
+            case double @double:
+                return $"{@double.ToString(DoubleFormatString, CultureInfo.InvariantCulture)}D";
+            case float.NegativeInfinity:
+                return "float.NegativeInfinity";
+            case float.PositiveInfinity:
+                return "float.PositiveInfinity";
+            case float.NaN:
+                return "float.NaN";
+            case float @float:
+                return $"{@float.ToString(SingleFormatString, CultureInfo.InvariantCulture)}F";
+            case decimal.MaxValue:
```

Comment:
Are the `MinValue`/`MaxValue` cases strictly necessary? The underlying numeric representations can still be used as literals and we're not special casing `MinValue`/`MaxValue` in the other numeric types.

Reply:
I think Eric copied the code from text.json which doing the same. But still a good question why we did that in the first place?

Reply:
Correct, I'm just moving this from JSON.  I believe this was covered here https://github.com/dotnet/roslyn/issues/58705 - C# actually doesn't like the numerical representation for these.

Reply:
> C# actually doesn't like the numerical representation for these.

It works so long as you slap an 'M' suffix at its end.

Reply:
I see - so probably my fix to use suffixes can remove this special case.